src/combinatorics/simple_graph.lean

```diff
@@ -0,0 +1,226 @@
+  v ∈ G.edge_of_adj h := sym2.mk_has_mem v w
+
+lemma mem_of_adj_right {v w : V} (h : G.adj v w) :
+  w ∈ G.edge_of_adj h := sym2.mk_has_mem_right v w
+
+lemma adj_iff_exists_edge {v w : V} (hne : v ≠ w) :
+G.adj v w ↔ ∃ (e : G.E), v ∈ e ∧ w ∈ e :=
+begin
+  split, { intro, use ⟦(v,w)⟧, assumption, refine ⟨(G.mem_of_adj _), (G.mem_of_adj_right _)⟩ },
+  rintro ⟨e, ⟨w', hve⟩, ⟨v', hew⟩⟩,
+  have : e.val = ⟦(v,w)⟧, { rw [hve, sym2.eq_iff] at hew ⊢, cc },
+  have key := e.property, rwa this at key,
+end
+
+variables {G}
+
+/-- Given an edge and one vertex incident on it, construct the other one. -/
+noncomputable def E.other (e : G.E) {v : V} (h : v ∈ e) : V :=
+by { have : v ∈ e.val, apply h, exact this.other }
+
+lemma E.other_mem (e : G.E) {v : V} (h : v ∈ e) : e.other h ∈ e :=
+begin
+  change sym2.mem.other h ∈ e.val, have := sym2.mem_other_spec h,
+  convert sym2.mk_has_mem_right _ _; tauto,
+end
+
+lemma E.other_ne (e : G.E) {v : V} (h : v ∈ e) : e.other h ≠ v :=
+begin
+  have key := e.property,
+  erw [← sym2.mem_other_spec h, sym2.eq_swap] at key,
+  exact G.ne_of_edge key,
+end
+
+attribute [irreducible] E.other
+variables (G)
+
+instance E.inhabited [inhabited {p : V × V | G.adj p.1 p.2}] : inhabited G.E :=
+⟨begin
+  rcases inhabited.default {p : V × V | G.adj p.1 p.2} with ⟨⟨x, y⟩, h⟩,
+  use ⟦(x, y)⟧, rwa sym2.from_rel_prop,
+end⟩
+
+instance edges_fintype [decidable_eq V] [fintype V] [decidable_rel G.adj] :
+  fintype G.E := subtype.fintype _
+
+attribute [irreducible] E
+
+@[simp] lemma irrefl {v : V} : ¬G.adj v v := G.loopless v
+
+@[symm] lemma edge_symm (u v : V) : G.adj u v ↔ G.adj v u := ⟨λ x, G.sym x, λ x, G.sym x⟩
+
+@[simp] lemma mem_neighbor_set (v w : V) : w ∈ G.neighbor_set v ↔ G.adj v w :=
+by tauto
+
+section finite_at
+
+/-!
+## Finiteness at a vertex
+
+This section contains definitions and lemmas concerning vertices that
+have finitely many adjacent vertices.  We denote this condition by
+`fintype (G.neighbor_set v)`.
+
+We define `G.neighbor_finset v` to be the `finset` version of `G.neighbor_set v`.
+Use `neighbor_finset_eq_filter` to rewrite this definition as a `filter`.
+-/
+
+variables (v : V) [fintype (G.neighbor_set v)]
+/--
+`G.neighbors v` is the `finset` version of `G.adj v` in case `G` is
+locally finite at `v`.
+-/
+def neighbor_finset : finset V := (G.neighbor_set v).to_finset
+
+@[simp] lemma mem_neighbor_finset (w : V) :
+  w ∈ G.neighbor_finset v ↔ G.adj v w :=
+by simp [neighbor_finset]
+
+/--
+`G.degree v` is the number of vertices adjacent to `v`.
+-/
+def degree : ℕ := (G.neighbor_finset v).card
+
+@[simp]
+lemma card_neighbor_set_eq_degree : fintype.card (G.neighbor_set v) = G.degree v :=
+by simp [degree, neighbor_finset]
+
+end finite_at
+
+section locally_finite
+
+variable [∀ (v : V), fintype (G.neighbor_set v)]
+
+/--
+A regular graph is a locally finite graph such that every vertex has the same degree.
+-/
+def regular_graph (d : ℕ) : Prop := ∀ (v : V), G.degree v = d
```

Comment:
```suggestion
A locally finite simple graph is regular of degree `d` if every vertex has degree `d`.
-/
def is_regular_of_degree (d : ℕ) : Prop := ∀ (v : V), G.degree v = d
```

Reply:
Not 100% sold on the new name suggestion, as it's a bit long. We do generally name `Prop`s `is_X` rather than just `X` though.

Reply:
in maths we might say is_d_regular, but i feel like we can't use `d` in the name if `d` appears only as a bound variable rather than as something typeful.

Reply:
actually I find `G.is_regular_of_degree d` pretty aesthetically nice, and it's not going to appear inside of longer terms very much.